File: leetcode48/leetcode48.py
# 第一种方法
class Solution:
    def rotate(self, matrix: list[list[int]]) -> None:
        """
        Do not return anything, modify matrix in-place instead.
        """
        # (i, j ) - > (j, n - 1 - i)
        tmp = copy.deepcopy(matrix)
        n = len(matrix)
        for i in range(n):
            for j in range(n):
                matrix[j][n - 1 - i] = tmp[i][j]


# 第二种方法
class Solution:
    def rotate(self, matrix: list[list[int]]) -> None:
        """
        Do not return anything, modify matrix in-place instead.
        """
        # (i, j ) - > (j, n - 1 - i)
        n = len(matrix)
        for i in range(n // 2):
            for j in range((n + 1) // 2):
                tmp = matrix[i][j]
                matrix[i][j] = matrix[n - 1 - j][i]
                matrix[n - 1 - j][i] = matrix[n - 1 - i][n - 1 - j]
                matrix[n - 1 - i][n - 1 - j] = matrix[j][n - 1 - i]
                matrix[j][n - 1 - i] = tmp

File: leetcode48/leetcode48.cpp
#include<bits/stdc++.h>
using namespace std;


// 第一种方法，使用辅助数组
class Solution {
public:
    void rotate(vector<vector<int>>& matrix) {
        // [i][j] -> [j][n - 1 - i]
        int n = matrix.size();
        vector<vector<int>>& temp = matrix;
        for (int i = 0; i < n; i++){
            for (int j = 0; j < n; j++){
                matrix[j][n - 1 - i] = temp[i][j];
            }
        }
    }
};

// 第二种方法，原地旋转
class Solution {
public:
    void rotate(vector<vector<int>>& matrix) {
        // [i][j] -> [j][n - 1 - i]
        int n = matrix.size();
        for (int i = 0; i < n / 2; i++){
            for (int j = 0; j < (n + 1) / 2; j++){
                int tmp = matrix[i][j];
                matrix[i][j] = matrix[n - 1 - j][i];
                matrix[n - 1 - j][i] = matrix[n - 1 - i][n - 1 - j];
                matrix[n - 1 - i][n - 1 - j] = matrix[j][n - 1 - i];
                matrix[j][n - 1 - i] = tmp;
            }
        }
    }
};